packages/react-spinbutton/Spec.md

```diff
@@ -2,62 +2,381 @@
+`SpinButton`'s `value` prop is always a number, in contrast to the v8 implementation that gave `value` a string typez. `SpinButton`s manipulate numeric values and making `value` a number aligns it with the other related props: `min`, `max` and `step`. `SpinButton`'s `value` is always displayed as a string which is determined with the `formatter()` function. By default this function stringifies `value` but it can do more sophisticated formatting like apply currency formatting. The inverse of `formatter()` is `parser()` which takes the stringified display value and converts it back to a number. The default implementation calls `parseFloat()` on the display value. `parser()` is called when a user directly types a value into `SpinButton`'s `<input>` element so this value can be converted to a number for computation. If `parser()` returns a non-number value (e.g., `NaN`, `undefined`, `null`) `value` will not be changed.
+
+`formatter()` and `parser()` allow users to hook into `SpinButton`'s display behavior without having to reimplement features like clamping `value` between `min` and `max` which is required when adding [custom suffixes in v8](https://codepen.io/seanms/pen/QWqpQWp).
+
+`SpinButton` does not allow values less than `min` or greater than `max` when these props are specified. When using the step buttons/arrow keys the value will not step outside the min/max bounds. If a value is typed into `<input>` that is outside the min/max range it will be clamped to the relevant bound, that is values less than `min` will be set to `min` and values greater than `max` will be set to `max`.
+
+Aside from min/max range clamping `SpinButton` does not currently implement any input validation.
+
+No error states are currently implemented.
+
+`SpinButton`'s `onChange` callback is invoked every time a change is committed. A change is committed when:
+
+1. A step button is activated
+2. When a user presses the `Up` or `Down` arrow keys while focused on the component.
+3. When a user presses the `Enter` key while focused on the `<input>`
+4. When the `blur` event is fired from the `<input>`
+
+The `onChange` callback is not invoked while a user is focused on the `<input>` and editing the value of the `SpinButton` directly.
+
+### Component States
 
-- _Component States_
-- _Interaction_
-  - _Keyboard_
-  - _Cursor_
-  - _Touch_
-  - _Screen readers_
+- Normal (no value or valid value, not focused)
+- Focused
+- Disabled
+- Error/Warning (value is invalid in some way). NB: error and validation are not currently supported.
+
+### Interaction
+
+#### Keyboard
+
+- Tab focuses the `<input>` element
+- Up arrow key increments the value by `step` until `max`, if specified.
+- Down arrow key decrements the value by `step` until `min`, if specified.
+- Home key sets the value to `min`, if specified.
+- End key sets the value to `max`, if specified.
+- Tab never focuses the increment or decrement buttons as their functionality is fully available via the keyboard
```

Comment:
ARIA spec also mentions `Page Up` and `Page Down`. Will be this implemented?

https://www.w3.org/TR/wai-aria-practices/examples/spinbutton/datepicker-spinbuttons.html

Reply:
Those interactions are optional and are very specific to what the spin button actually represents with numbers. In the case of a calendar it could be per week/month but we shouldn't implement this in our component since it's so scenario specific

Reply:
I think the linked example is specific to date picker spin buttons, which is not what we're implementing here. This would be the relevant general section. https://w3c.github.io/aria-practices/#keyboard-interaction-18

Reply:
Since they are optional in the spec I intentionally left them off to minimize `SpinButton`'s API surface. We can add them later if we need them.